connections/multiselect2.md

```diff
@@ -0,0 +1,103 @@
+# Multiselect 2.0
+
+## Introduction
+
+Multiselect 2.0 replaces the Multistream protocol. Compared to its predecessor, it offers:
+
+1. Downgrade protection for the security protocol negotiation.
+2. Zero-rountrip stream multiplexer negotiation for handshake protocols that take advantage of early data mechanisms (one-roundtrip negotation for protocols / implementations that don't).
+3. Compression for the protocol identifiers of frequently used protocols.
+
+By using protobufs for all control messages, Multiselect 2.0 provides an easy path for future protocol upgrades. The protobuf format guarantees that unknown field in a message will be skipped, thus future version of the protocol can add new fields that signal support for new protocol features.
+
+## High-Level Overview
+
+### Handshake Protocol Selection
+
+Handshake protocols are not being negotiated, but are announced in the peers' multiaddrs. 
+
+**TODO**: Do we need to describe the format here? I guess we don't, but we will probably need another document for that change, and we can link to it from here.
+
+Peers advertising a multiaddr that includes a handshake protocol MUST support Multiselect 2.0 as described in this document.
+
+#### TCP Simultaneous Open
+
+TCP allows the establishment of a connection if two endpoints start initiating a connection at the same time. This is called TCP Simultaneous Open. For libp2p, this is problematic, since most stream multiplexers assign stream IDs based on the role (client or server) of and endpoint.
+
+It is therefore desirable to fail a connection attempt as early as possible, if a TCP Simultaneous Open occurs. TLS 1.3 and Noise provide this guarantee: For example, the TLS handshake will fail if an endpoint receives a ClientHello instead of a ServerHello as a response to its ClientHello.
```

Comment:
@vyzo, is it true that we want to fail the connection completely if a TCP simultaneous open occurs? I remember you were thinking though how to do NAT hole-punching with TCP and that simultaneous open was part of that story. 

It seems like for hole punching, we want one of the simultaneous connection attempts to succeed, but we need a deterministic way to figure out which party has the initiator role. I don't know how that fits in to this multiselect design, since multiselect now only kicks in _after_  the crypto handshake, and it seems like TLS and Noise would both just fail the handshake if both parties believe they are the initiator.

Reply:
One of the requirements of ms2.0 is to be deep-packet inspection censorship resistant. This means not exchanging easily recognisable traffic in plaintext (e.g. secure channel negotiation, or tokens like `iamclient`).

We now advertise secure channels in the multiaddr, and the initiator jumps right into the handshake.

IMO we need a conflict resolution mechanism that sits above multiselect 2.0, detects TCP simultaneous connects, and resolves the conflict via some deterministic or probabilistic heuristic.

An example of a deterministic heuristic could be:

> If we are the initiator and we receive an invalid payload from the responder (plus points if we can detect it was a ClientHello – TLS 1.3 – or the wrong Noise message), both sides hash the outgoing bytes and incoming bytes exchanged until then, and the highest value becomes the initiator. The handshake is tried again with each side taking their new role.

That scheme would require us to "tee" the `io.Reader` and `io.Writer` until the handshake is over, which is probably fine.

An example of a probabilistic heuristic could be:

> If we are the initiator and receive an invalid payload from the responder, we roll a dice to determine if we're initiator or responder, and try again, up to a maximum of 3 times.